src/MDOP.jl

```diff
@@ -3,37 +3,154 @@ using StatsBase
+considered delay values is determined in `τs` and for the nearest neighbor
+search we respect the Theiler window `w`.
+
+After each embedding cycle the FNN-statistic `FNNs` [^Hegger1999][^Kennel1992] is being
+checked and as soon as this statistic drops below the threshold `fnn_thres`,
+the algorithm terminates. In order to increase the practability of the method the
+algorithm also terminates when the FNN-statistic `FNNs` increases.
+
+The final embedding is returned as `Y`. The chosen delay values for each embedding
+cycle are stored in the `τ_vals` and the according timeseries index chosen for the
+the respective according delay value in `τ_vals` is stored in `ts_vals`.
+`βS, FNNs` are returned for clarity and double-checking, since they are computed anyway.
 
 [^Nichkawde2013]: Nichkawde, Chetan (2013). [Optimal state-space reconstruction using derivatives on projected manifold. Physical Review E 87, 022905](https://doi.org/10.1103/PhysRevE.87.022905).
+[^Hegger1999]: Hegger, Rainer and Kantz, Holger (1999). [Improved false nearest neighbor method to detect determinism in time series data. Physical Review E 60, 4970](https://doi.org/10.1103/PhysRevE.60.4970).
+[^Kennel1992]: Kennel, M. B., Brown, R., Abarbanel, H. D. I. (1992). [Determining embedding dimension for state-space reconstruction using a geometrical construction. Phys. Rev. A 45, 3403] (https://doi.org/10.1103/PhysRevA.45.3403).
 """
-function MDOP(Y...)
-    #tbd
+function mdop_embedding(s::Vector{T};
+        τs = 0:50 , w::Int = 1, fnn_thres::Real = 0.05,
+        max_num_of_cycles = 50,
+        r::Real = 2) where {T<:Real}
+
+    @assert 0 ≤ fnn_thres < 1 "Please select a valid breaking criterion, `fnn_thres` ∈ [0 1)"
+    @assert all(x -> x ≥ 0, τs)
+    metric = Euclidean()
+    s = regularize(s) # especially important for fnn-computation
+    # define actual phase space trajectory
+    Y_act = Dataset(s)
+
+    # compute nearest neighbor distances in the first embedding dimension for
+    # FNN statistic
+    vtree = KDTree(Y_act, metric)
+    allNNidxs, NNdist_old = all_neighbors(vtree, Y_act, 1:length(Y_act), 1, w)
+
+    # set a flag, in order to tell the while loop when to stop. Each loop
+    # stands for encountering a new embedding dimension
+    flag, counter = true, 1
 
+    τ_vals = Int64[0]
+    ts_vals = Int64[1]
+    FNNs = Float64[]
+    βS = Array{T}(undef, length(τs), max_num_of_cycles)
+
+    # loop over increasing embedding dimensions until some break criterion will
+    # tell the loop to stop/break
+    while flag
+        Y_act, NNdist_new = mdop_embedding_cycle!(
+            Y_act, flag, s, τs, w, counter, βS, τ_vals, metric, r,
+            FNNs, fnn_thres, ts_vals, NNdist_old
+        )
+        flag = mdop_break_criterion(FNNs, counter, fnn_thres, max_num_of_cycles)
+        counter += 1
+        # for the next embedding cycle save the distances of NN for this dimension
+        NNdist_old = NNdist_new
```

Comment:
> I see that beta_statistic calculates all_neighbors, but this calculation has already been done for its input argument Y. Why do we have to re-calculate the neighbors instead of just passing the calculation along? Getting the nearest neighbors is the most costly operation in this function, it should be done only when absolutely necessary.

That is just partly correct. Indeed `beta_statistic` computes all neighbours, and is has to. For the computations of FNN's in each embedding cycle `fnn_embedding_cycle` needs all NN-distances for the present and all NN-distances for the corresponding next embedding dimension. `beta_statistic` computes, thus, the "new" NN-distances in each cycle and these distances are then being used for the next cycle as the "old" NN-distances (see marked code), so it is all good. Just for the very first iteration cycle (==1st embedding cycle) the "old" distances do not exist so we compute them in line 69.